### PR TITLE
Quirk waits for client before setting preferences

### PR DESCRIPTION
**Pull request by nightred**
## About The Pull Request

Quirks used a timer to run post_add to set the client preferences.
We will now validate that we have a client before applying preferences.

This was not an issue on local servers but it can happen on the live servers more often.

## Why It's Good For The Game

Should solve the random phobia when spawn takes longer then normal.

## Changelog
:cl:
fix: Quirks settings trigger after spawn now.
/:cl:

### CODE REVIEW

File: code/datums/traits/_quirk.dm [hunk 1]
```diff
@@ -12,6 +12,7 @@
 	var/mob_trait //if applicable, apply and remove this mob trait
 	///Amount of points this trait is worth towards the hardcore character mode; minus points implies a positive quirk, positive means its hard. This is used to pick the quirks assigned to a hardcore character. 0 means its not available to hardcore draws.
 	var/hardcore_value = 0
+	var/post_add_ran = FALSE /// Tracks if the post_add was run on the quirk on process tick
```
Comment (Rohesie) [inline code/datums/traits/_quirk.dm:15]: Could you turn this into a bitflag? Something like `var/quirk_flags = NONE`, `#define QUIRK_POST_ADD_RAN (1<<0)` and `if(!(quirk_flags & QUIRK_POST_ADD_RAN))`

File: code/datums/traits/_quirk.dm [hunk 2]
```diff
@@ -65,6 +65,13 @@
 		return
 	if(quirk_holder.stat == DEAD)
 		return
+	// post_add is used to get settings from the client prefs
+	// validate that a client is attached before running
+	if(!post_add_ran)
+		var/mob/user = quirk_holder
+		if(user && user.client)
```
Comment (Rohesie) [inline code/datums/traits/_quirk.dm:72]: `quirk_holder` is already typed as living and prior checks imply it exists at this point. Just do something like `if(!(quirk_flags & QUIRK_POST_ADD_RAN) && quirk_holder.client)`

File: code/datums/traits/_quirk.dm (hunk 1 above)
Comment (nightred) [reply code/datums/traits/_quirk.dm:15]: Making this change.

File: code/datums/traits/_quirk.dm (hunk 2 above)
Comment (nightred) [reply code/datums/traits/_quirk.dm]: Fixed this up also.

**Comment by Rohesie** [comment]
Thinking about this better, you shouldn't repeat the same check every process.
Instead what you want to do is check, when `quirk_holder` is added, if it has a client.
If it has, run the proc, else, register to `COMSIG_MOB_LOGIN`, and do your thing then, running the proc and unregistering.

**Comment by nightred** [comment]
That signal is what I wanted and could not find.

File: code/datums/traits/_quirk.dm [hunk 3]
```diff
@@ -29,9 +29,10 @@
 	add()
 	if(spawn_effects)
 		on_spawn()
-		addtimer(CALLBACK(src, .proc/post_add), 30)
+	RegisterSignal(quirk_holder, COMSIG_MOB_LOGIN, .proc/post_add)
 
 /datum/quirk/Destroy()
+	UnregisterSignal(quirk_holder, COMSIG_MOB_LOGIN)
```
Comment (Rohesie) [inline code/datums/traits/_quirk.dm:35]:
```suggestion
```
Signals clear themselves properly already on deletion of either listener or emitter.

File: code/datums/traits/_quirk.dm [hunk 4]
```diff
@@ -29,9 +29,10 @@
 	add()
 	if(spawn_effects)
 		on_spawn()
-		addtimer(CALLBACK(src, .proc/post_add), 30)
+	RegisterSignal(quirk_holder, COMSIG_MOB_LOGIN, .proc/post_add)
```
Comment (Rohesie) [inline code/datums/traits/_quirk.dm:32]:
```suggestion
	if(quirk_holder.client)
		post_add()
	else
		RegisterSignal(quirk_holder, COMSIG_MOB_LOGIN, .proc/on_quirk_holder_first_login)


/datum/quirk/proc/on_quirk_holder_first_login(mob/living/source)
		UnregisterSignal(source, COMSIG_MOB_LOGIN)
		post_add()
```

File: code/datums/traits/_quirk.dm [hunk 5]
```diff
@@ -29,7 +29,16 @@
...
 	if(spawn_effects)
 		on_spawn()
-		addtimer(CALLBACK(src, .proc/post_add), 30)
+	if(quirk_holder.client)
+		post_add()
+	else
+		RegisterSignal(quirk_holder, COMSIG_MOB_LOGIN, .proc/on_quirk_holder_first_login)
+
+
+///Description of what this proc does here.
```
Comment (Rohesie) [inline code/datums/traits/_quirk.dm:38]: Do give the proc a proper documentation.
And ask if you don't understand how any of this works.
The idea is that you learn here how to make the code respond to events so it only runs a single time, not for me to just code it for you.

Comment (nightred) [reply code/datums/traits/_quirk.dm]: The idea of unregistering it after trigger is great i did not consider it.
I was approaching this as always later to add client i never thought they could be attached at this point.

I am happy for the feedback

File: code/datums/traits/_quirk.dm [hunk 6]
```diff
@@ -29,7 +29,16 @@
...
 	if(spawn_effects)
 		on_spawn()
-		addtimer(CALLBACK(src, .proc/post_add), 30)
+	if(quirk_holder.client)
+		post_add()
+	else
+		RegisterSignal(quirk_holder, COMSIG_MOB_LOGIN, .proc/on_quirk_holder_first_login)
+
+
+/// Run post spawn text and client preferences
```
Comment (Rohesie) [inline code/datums/traits/_quirk.dm:38]:
```suggestion
///Run post spawn text and client preferences.
```
No space between the 3 `/` and the comment, else it won't be properly documented.

Comment (nightred) [reply code/datums/traits/_quirk.dm:38]: really? that is a thing? *waves bye to the space*

Comment (nightred) [reply code/datums/traits/_quirk.dm]: Block comments for style.

**Pull request merged**